### PR TITLE
Add a policy for async communication

### PR DESCRIPTION
**Pull request by Flamefire**
This makes the asyncCommunication a functor instead of a virtual method.
This has the advantage (besides beeing more TMP like) that a call preserves the actual type compared to "concreteParticleInstance->asyncCommunication(event)" where functors down the chain only have access to ParticleBase.

As requested by @psychocoderHPC the functor can be specialized for any type. Particles are the only one used right now, but a specialization can be defined for buffers too without much effort.

For inheritance detection a tag is used ->"SimulationDataTag" type member. This is similar to e.g. iterator_tag from STL and solves the problem, that one cannot detect, whether a type is derived from a templated type.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,155 @@
+cmake_minimum_required(VERSION 2.8.12.2)
```
Comment (psychocoderHPC) [inline src/CMakeLists.txt]: might be this fie should not be included in this pull request

File: src/picongpu/include/particles/ParticlesFunctors.hpp [hunk 2]
```diff
@@ -184,10 +185,9 @@ struct CommunicateSpecies
         if (hasPusher::value)
         {
             EventTask updateEvent(*(updateEventList.begin()));
-            PMACC_AUTO(speciesPtr, tuple[SpeciesName()]);
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ParticlesFunctors.hpp]: please revert this line and use the interim variable

File: src/libPMacc/include/particles/AsyncCommunicationImpl.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2015 Alexander Grund
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/AsyncCommunicationImpl.hpp]: This list should contain all authors of src/libPMacc/include/particles/ParticlesBase.tpp because most of the file is copied.

File: src/libPMacc/include/communication/AsyncCommunication.hpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+     *     struct AsyncCommunicationImpl<
+     *         T_Data,
+     *         Bool2Type< boost::is_integral<T_Data>::value >
+     *     >{...}
+     */
+    template<typename T_Data, typename T_IsSpecialized = Bool2Type<true> >
+    struct AsyncCommunicationImpl;
+
+    /**
+     * This policy starts an asynchronous communication of the given data
```
Comment (psychocoderHPC) [inline src/libPMacc/include/communication/AsyncCommunication.hpp]: thanks for the nice documentation in all new functions!

File: src/libPMacc/include/communication/AsyncCommunication.hpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+     * Implementations of \see AsyncCommunication should specialize this,
+     * but it is not intended to be called directly. Use \see AsyncCommunication
+     *
+     * The 2nd template parameter can be used to check for conditions on
+     * templated implementations. E.g.:
+     *
+     *     template<typename T_Data>
+     *     struct AsyncCommunicationImpl<
+     *         T_Data,
+     *         Bool2Type< boost::is_integral<T_Data>::value >
```
Comment (psychocoderHPC) [inline src/libPMacc/include/communication/AsyncCommunication.hpp]: one `>` at the end is missing

Comment (Flamefire) [inline src/libPMacc/include/communication/AsyncCommunication.hpp]: No, it's on the next line

**Comment by psychocoderHPC** [comment]
please run KHI heating and check the macro particles count

**Comment by Flamefire** [comment]
Both tests succeeded
